test/test_nn.py

```diff
@@ -2505,6 +2526,131 @@ def test_embedding_bag(self):
         self._test_EmbeddingBag(False, 'sum', True)
         self._test_EmbeddingBag(False, 'mean', True)
 
+    @staticmethod
+    def _get_reduction(mode='mean', dim=0):
+        def safe_mean(tensor, dim):
+            if tensor.size(dim) > 0:
+                return tensor.mean(dim)
+            return torch.zeros_like(tensor.mean(dim))
+
+        def safe_max(tensor, dim):
+            if tensor.size(dim) > 0:
+                return tensor.max(dim)[0]
+            return torch.zeros_like(tensor.mean(dim))
+
+        if mode == 'mean':
+            return lambda t: safe_mean(t, dim)
+        elif mode == 'sum':
+            return lambda t: t.sum(dim)
+        elif mode == 'max':
+            return lambda t: safe_max(t, dim)
+        else:
+            assert false
+
+    @staticmethod
+    def _embedding_bag_reference(input, weight, offsets=None, mode='mean',
+                                 per_sample_weights=None):
+        assert offsets is not None
+        if per_sample_weights is None:
+            per_sample_weights = torch.ones(input.size())
+        assert input.numel() == per_sample_weights.numel()
+
+        reduction = TestNN._get_reduction(mode)
+        bags = []
+        embeddings = weight.index_select(0, input) * per_sample_weights.unsqueeze(1)
+        for index, offset in enumerate(offsets):
+            if index + 1 < len(offsets):
+                next_offset = offsets[index + 1]
+            else:
+                next_offset = len(input)
+            length = next_offset - offset
+            bags.append(reduction(embeddings.narrow(0, offset, length)))
+        return torch.stack(bags)
+
+    @staticmethod
+    def _test_EmbeddingBag_per_sample_weights_failures(self, device='cpu'):
```

Comment:
It might be worth adding a dtype kwarg to this and to test the EmbeddingBag with weights set to both float and double.

Reply:
Sure

Reply:
Actually, it doesn't matter for this case: adding the extra test would be redundant IMO. This case is checking if EmbeddingBag throws the correct errors on mismatched inputs. The float/double code use the same codepath until we get to the actual computation (which is tested in other tests but not this one)